rosbag2_transport/include/rosbag2_transport/player_service_client.hpp

```diff
@@ -0,0 +1,99 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef ROSBAG2_TRANSPORT__PLAYER_SERVICE_CLIENT_HPP_
+#define ROSBAG2_TRANSPORT__PLAYER_SERVICE_CLIENT_HPP_
+
+#include <chrono>
+#include <map>
+#include <memory>
+#include <mutex>
+#include <string>
+
+#include "rclcpp/generic_client.hpp"
+#include "rclcpp/rclcpp.hpp"
+
+namespace rosbag2_transport
+{
+
+class PlayerServiceClientManager;
+
+class PlayerServiceClient final
+{
+public:
+  explicit PlayerServiceClient(
+    std::shared_ptr<rclcpp::GenericClient> cli,
+    std::string service_name,
+    const std::string & service_event_type,
+    const rclcpp::Logger logger,
+    std::shared_ptr<PlayerServiceClientManager> player_service_client_manager);
+
+  // Can call this function if check_include_request_message() return true
+  void async_send_request(const rclcpp::SerializedMessage & message);
+
+  std::shared_ptr<rclcpp::GenericClient> generic_client()
+  {
+    return client_;
+  }
+
+  // Check if message can be unpacked to get request message
+  bool is_include_request_message(const rclcpp::SerializedMessage & message);
+
+private:
+  std::shared_ptr<rclcpp::GenericClient> client_;
+  std::string service_name_;
+  const rclcpp::Logger logger_;
+  std::shared_ptr<PlayerServiceClientManager> player_service_client_manager_;
+  enum class introspection_type
+  {
+    UNKNOW = 0,
+    METADATA,
+    CONTENTS
+  };
+  introspection_type client_side_type_ = introspection_type::UNKNOW;
+  introspection_type service_side_type_ = introspection_type::UNKNOW;
+
+  std::shared_ptr<rcpputils::SharedLibrary> ts_lib_;
+  const rosidl_message_type_support_t * ts_;
+  const rosidl_typesupport_introspection_cpp::MessageMembers * message_members_;
+
+  rcutils_allocator_t allocator_ = rcutils_get_default_allocator();
+
+  uint8_t get_msg_event_type(const rclcpp::SerializedMessage & message);
+};
+
+class PlayerServiceClientManager final
+{
+public:
+  PlayerServiceClientManager(
+    size_t request_future_timeout, size_t maximum_request_future_queue = 30);
+
+  // Timeout future will be discarded and check queue.
+  bool request_future_queue_is_full();
+
+  bool register_request_future(rclcpp::GenericClient::FutureAndRequestId & request_future);
+
+private:
+  using time_point = std::chrono::steady_clock::time_point;
+  using ptr_future_and_request_id = std::unique_ptr<rclcpp::GenericClient::FutureAndRequestId>;
+  std::map<time_point, ptr_future_and_request_id> request_futures_list_;
+  std::mutex request_futures_list_lock_;
+
+  std::chrono::seconds request_future_timeout_;
+  size_t maximum_request_future_queue_;
```

Comment:
I do believe those are really user dependent configuration, that would be really better to configure when we play the service bag as clients? what do you think?

@MichaelOrlov @Barry-Xu-2018

Reply:
furthermore, if the rosbag2 plays back with really low rate, or server is working with `use_sim_time` enabled, eventually server would take really long time to respond to the rosbag2 service client for this. that means, user cannot receive the service response from the server. i think this needs to be well explained at least when we use `ros2 bag play services`. at least, these options should be configurable by user, and they understand that this timeout is based on monotonic clock.

Reply:
> I do believe those are really user dependent configuration, that would be really better to configure when we play the service bag as clients? what do you think?

Totally agree.    
Maybe request_future_timeout should be configured by user. We also don't know regarding the required duration of a service process.  Currently, it is set for 5 minutes, which should suffice for some cases. However, for cases with special requirements, we should also provide a configurable method.  

At the beginning, Michael suggested this way https://github.com/ros2/rosbag2/pull/1481#issuecomment-1774755277. I used another way and wanted to hear opinions. So I neglected to consider allowing users to set the relevant parameters.

Reply:
hmm, having async behavior looks good to me as implemented now with max queue depth and timeout reap the pending request if that takes long time. if we have thread to do sync way, we might have thousands threads running to wait the response from service server, we have no idea how much it takes to complete the service.

i will defer this to @MichaelOrlov , what do you think?